From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Wed, 18 Mar 2015 22:24:08 +0800
Subject: [PATCH] {MAYBE UPS} include/linux/jiffies: fix inconsistencies for
 jiffies section

[TODO] This is not upstreamable as-is because some
architectures define jiffies = jiffies_64 + 4, which is obviosly not aligned.
But it is still in the same section... Find out if the annotations only
define the section or if they are stronger, and also if it is okay to mix
__cacheline_aligned and __cacheline_aligned_in_smp...

On x86_64, jiffies is declared in section .data in
include/linux/jiffies.h but in section .data..cacheline_aligned (through
__cacheline_aligned attribute) in arch/x86/kernel/time.c.  Fix this
inconsistency by forcing jiffies in .data..cacheline_aligned in the
header.

Since commit ecea8d19c9f0 ("[PATCH] jiffies_64 cleanup"), jiffies_64 is
defined in kernel/time/timer.c with __cacheline_aligned_in_smp.  Use the
same marker in its definition in include/linux/jiffies.h to fix a
section mismatch between a declaration in .data and a definition in
.data..cacheline_aligned.

[update]
Since commit 7c30f352c852 ("jiffies.h: declare jiffies and jiffies_64
with ____cacheline_aligned_in_smp"), jiffies_64 is rightly in
__cacheline_aligned_in_smp but there is still a section mismatch issue
with jiffies.  Continue using __cacheline_aligned on x86_64.

Signed-off-by: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
---
 include/linux/jiffies.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/linux/jiffies.h b/include/linux/jiffies.h
index a27cf6652327..ab0ed4493780 100644
--- a/include/linux/jiffies.h
+++ b/include/linux/jiffies.h
@@ -78,7 +78,11 @@ extern int register_refined_jiffies(long clock_tick_rate);
  * get_jiffies_64() will do this for you as appropriate.
  */
 extern u64 __cacheline_aligned_in_smp jiffies_64;
+#ifdef __x86_64__
+extern unsigned long volatile __cacheline_aligned jiffies;
+#else
 extern unsigned long volatile __cacheline_aligned_in_smp __jiffy_arch_data jiffies;
+#endif
 
 #if (BITS_PER_LONG < 64)
 u64 get_jiffies_64(void);
-- 
